lib/core/test/unit/Cardano/Pool/MetadataSpec.hs

```diff
@@ -0,0 +1,239 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TypeApplications #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Cardano.Pool.MetadataSpec
+    ( spec
+    ) where
+
+import Prelude
+
+import Cardano.Pool.Metadata
+    ( Api
+    , BaseUrl (..)
+    , Client (..)
+    , Scheme (..)
+    , defaultManagerSettings
+    , mkClient
+    , newManager
+    )
+import Cardano.Wallet.Api.Server
+    ( Listen (..), withListeningSocket )
+import Cardano.Wallet.Api.Types
+    ( ApiT (..) )
+import Cardano.Wallet.Primitive.Types
+    ( PoolId (..), StakePoolOffChainMetadata (..), StakePoolTicker (..) )
+import Control.Concurrent.Async
+    ( async, cancel )
+import Control.Concurrent.MVar
+    ( newEmptyMVar, putMVar, takeMVar )
+import Control.Exception
+    ( bracket )
+import Control.Lens
+    ( at, (.~), (?~) )
+import Data.Aeson
+    ( toJSON )
+import Data.Function
+    ( (&) )
+import Data.Proxy
+    ( Proxy (..) )
+import Data.String
+    ( fromString )
+import Data.Swagger
+    ( Referenced (..)
+    , Schema
+    , SwaggerType (..)
+    , maxLength
+    , minLength
+    , paramSchema
+    , properties
+    , required
+    , type_
+    )
+import Data.Swagger.Schema.Validation
+    ( validateJSON )
+import Data.Text
+    ( Text )
+import Data.Text.Class
+    ( ToText (..) )
+import Data.Vector.Shuffle
+    ( mkSeed )
+import Network.HTTP.Types
+    ( status404 )
+import Network.Wai.Handler.Warp
+    ( runSettingsSocket, setBeforeMainLoop )
+import Servant
+    ( Server, err404, serve, throwError )
+import Servant.Client
+    ( ClientError (..), responseStatusCode )
+import Test.Hspec
+    ( Spec, around, describe, it )
+import Test.QuickCheck
+    ( ASCIIString (..)
+    , Arbitrary (..)
+    , PrintableString (..)
+    , choose
+    , counterexample
+    , label
+    , property
+    , vector
+    , vectorOf
+    )
+import Test.QuickCheck.Gen
+    ( Gen (..) )
+import Test.QuickCheck.Monadic
+    ( assert, monadicIO, monitor, run )
+import Test.QuickCheck.Random
+    ( mkQCGen )
+
+import qualified Data.Aeson as Aeson
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy.Char8 as BL8
+import qualified Data.Text as T
+import qualified Network.Wai.Handler.Warp as Warp
+
+spec :: Spec
+spec = describe "Metadata - MockServer" $
+    around withMockServer $ it "Mock Server works as intended" $
+        \Client{getStakePoolMetadata} -> property $ \pid -> monadicIO $ do
+            run (getStakePoolMetadata pid) >>= \case
+                Left (FailureResponse _ res) -> do
+                    monitor $ label "No Corresponding Metadata"
+                    monitor $ counterexample $ mconcat
+                        [ show (responseStatusCode res)
+                        , " =/= "
+                        , show status404
+                        ]
+                    assert (responseStatusCode res == status404)
```

Comment:
Might be nicer to use something like [`assertWith`](https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/test/unit/Cardano/Wallet/DB/Properties.hs#L391-L396) to couple the assertion with the error message

Now the `counterexample` is dependent on the next line.

… though perhaps not worth for just one instance.

Reply:
Yeah, I wish `assertWith` was part of QuickCheck though. I might as well do a PR actually :)

Reply:
https://github.com/nick8325/quickcheck/pull/294 :man_shrugging: